Full Static Control (Except Invoke) Inlining & Compilation

calebmkim opened the pull request:
Adds static `seq`, `par`, `if`, and `repeat` support, in addition to static `enable`!  

Currently, it inlines each of these four different control operators into a single static group in the `static-inline` pass, and then the `compile-static` pass 1) instantiates FSMs so that these inlined static groups get turned into dynamic groups with "early-resetting" FSMs that never reach the group's done condition, and 2) creates wrapper groups for these "early reseting" groups when necessary (i.e., when they need to use a go-done interface). 

For the correctness test cases, it passes the same test cases as the `tdst` pass did (only difference is that it doesn't infer static control yet, but we adding that support rn) . I also added another case that compiles a "static island" within dynamic control just to demonstrate that we can do that as well. 

A Couple Notes: 
- Because of the recursive nature of the static-inlining pass, it ends up creating a bunch of unnecessary timing guards for nested control statements (e.g., `%3 & %[2:4] & %[0:5]`, which is really just `%3`). I'll write up an issue on this... ideally, we want some sort of `simplify-guards` pass that can handle this separately. 
- StaticInvoke support is still a to-do (#1417).  

When merged, should close #1406,. After this, we can also delete `merge-static-par` and `static-par-conv` as these passes no longer really do anything.

rachitnigam commented:
Great! I don't have time to review this yet but make sure that the programs take exactly the same amount of time (in number of cycles) as the `tdst` programs did. The "dynamic island" test is a great idea and something I wanted to do with the OG `tdst` pass but it couldn't support it. Great to see it become a reality.

Also, I mentioned in #1419, we should have a new alias for the compilation pipeline that uses the new `static` powered passes and eventually remove all the old crud.

rachitnigam commented on calyx-frontend/src/ast.rs:
```diff
@@ -341,6 +343,54 @@ pub enum Control {
         attributes: Attributes,
     },
+    /// Represents sequential composition of static control statements.
+    StaticSeq {
```
One other way to do this is having an enum like this:
```rust
enum ControlType {
  Dynamic,
  Static { latency: u64 }
}
```

And then make all the enum variants for the `Control` enum specify a `ControlType`. This probably doesn't matter too much since the frontend is not often used by any other part of the compiler library.

EclecticGriffin replied:
Possibly worth doing, but I will also note that while redundant, this makes it possible to write code that knows at a type level that it will only be working with static groups, rather than having to check the `ControlType` variant and error out in certain cases. Not sure if this is worth the repetition but I figure it's worth mentioning.

calebmkim replied:
Yeah, when going from AST -> IR, I actually use this idea to give an error when someone tries to put dynamic control in static context, so I'll prob keep as is. Not that it really matters either way, though.

rachitnigam commented on calyx-frontend/src/parser.rs:
```diff
@@ -871,16 +932,32 @@ impl CalyxParser {
             [enable(data)] => data,
             [empty(data)] => data,
             [invoke(data)] => data,
             [seq(data)] => data,
```
The benefit of the approach for the `enum` would be that you don't need to have two different parsing variants for each one of the control operators and can maybe parse them in one function

rachitnigam commented on calyx-ir/src/control.rs:
```diff
@@ -477,6 +479,11 @@ impl Control {
         })
     }
 
+    /// Convience constructor for if
```
Incorrect comment

rachitnigam commented on calyx-ir/src/from_ast.rs:
```diff
@@ -568,12 +568,160 @@ fn build_static_guard(
+            comp: component,
+            attributes,
+        } => match builder.component.find_group(component) {
+            Some(_) => {
```
The `match` can probably be replaced with a conditional for the erroneous path:
```rust
if builder.component.find_group(component).is_some() {
  return Err(...)
}
// Normal case
```

In general avoid nesting using `match` when possible

rachitnigam commented on calyx-ir/src/guard.rs:
```diff
@@ -79,6 +79,10 @@ impl StaticTiming {
         self.interval
     }
+
+    pub fn set_interval(&mut self, new_interval: (u64, u64)) {
```
Doc comments needed

rachitnigam commented on calyx-opt/src/passes/compile_static.rs:
```diff
@@ -75,99 +87,265 @@ fn make_guard_dyn(
     ir::Assignment {
         src: assign.src,
-        dst: new_dst,
+        dst: assign.dst,
```
Often times when you're replacing some fields of a struct, you can use the following syntax:
```
ir::Assignment {
  guard: make_guard_dyn(...)
  ..assign
}
```

The `..assign` tells rust to replace all the other fields. This might not work here because you're building an `ir::Assignment<Nothing>` here from a `ir::Assignment<StaticTiming>`.

calebmkim replied:
yeah, doesn't work cuz of type differences unfortunately.

rachitnigam commented on calyx-opt/src/passes/compile_static.rs:
```diff
@@ -75,99 +87,265 @@ fn make_guard_dyn(
+            .find(|port| port.borrow().name == "in")
+            .unwrap_or_else(|| panic!("called {} in build_wrapper_group as an fsm; should have `in` port", early_reset_fsm.borrow().name()))
+            .borrow()
+            .width;
```
This is probably not the best way to get the width of the FSM. An easier solution would be to pass the width to the function. Alternatively, you can look at the parameters of the parent cell of the FSM instead of looking at the width of the `in` port.

rachitnigam commented on calyx-opt/src/passes/compile_static.rs:
```diff
@@ -75,99 +87,265 @@ fn make_guard_dyn(
+            .get_groups()
+            .find(*group_name)
+            .unwrap_or_else(|| {
+                panic!(
```
Per @EclecticGriffin's recommendation, avoid using `panic!` directly. If this case should never occur, use `unreachable!` instead

rachitnigam commented on calyx-opt/src/passes/compile_static.rs:
```diff
@@ -75,99 +87,265 @@ fn make_guard_dyn(
+            });
+        let early_reset_fsm =
+            builder.component.find_cell(*fsm_name).unwrap_or_else(|| {
+                panic!(
```
Same comment about `panic!`

rachitnigam commented on calyx-opt/src/passes/compile_static.rs:
```diff
@@ -75,99 +87,265 @@ fn make_guard_dyn(
+                *e.get_mut_attributes() = attrs;
+                Ok(Action::Change(Box::new(e)))
+            }
+            _ => unreachable!("Non-Enable Static Control should have been compiled away. Run static-inliner to do this"),
```
Don't hard code the name of the pass. Instead, import it and use the `::name()` method on it to generate the correct pass name.

rachitnigam commented on calyx-opt/src/passes/compile_static.rs:
```diff
@@ -75,99 +87,265 @@ fn make_guard_dyn(
+    ) -> VisResult {
+        // assume that there are only static enables left.
+        // if there are any other type of static control, then error out.
+        match sc {
```
Might be the right place to use the new [`let`-`else` construct](https://rust-lang.github.io/rfcs/3137-let-else.html) in Rust:
```
let ir::StaticControl::Enable(s) = sc else {
  unreachable!(...)
};
// Remaning stuff that uses `s`
```

rachitnigam commented on calyx-opt/src/traversal/visitor.rs:
```diff
@@ -319,6 +319,28 @@ pub trait Visitor {
     }
 
+    /// Executed after visiting the children of a [ir::While] node.
+    fn start_static_control(
```
Incorrect comment

rachitnigam commented on calyx-opt/src/traversal/visitor.rs:
```diff
@@ -319,6 +319,28 @@ pub trait Visitor {
+        Ok(Action::Continue)
+    }
+
+    /// Executed after visiting the children of a [ir::While] node.
```
Incorrect comment

rachitnigam commented on tests/correctness/static-control/static-island.futil:
```diff
@@ -0,0 +1,49 @@
+            if lt_cond.out {
+                add_five_reg; 
+            }
+            static seq <16> {
```
Is it the case that the program has to explicitly specify the latency for each control operator at the use point? For example, can a user instead provide `static seq { ... }` and have the compiler automatically infer the right latency?

rachitnigam commented on calyx-opt/src/passes/static_inliner.rs:
```diff
@@ -0,0 +1,388 @@
+
+impl Named for StaticInliner {
+    fn name() -> &'static str {
+        "static-inline"
```
This is why I recommend using the `::name` function. The error message above says `static-inliner` but the pass is called `static-inline`.

rachitnigam commented on calyx-opt/src/passes/static_inliner.rs:
```diff
@@ -0,0 +1,388 @@
+    fn make_cond_assigns(
+        cond: &ir::RRC<ir::Cell>,
+        cond_wire: &ir::RRC<ir::Cell>,
+        port: &ir::RRC<ir::Port>,
```
I generally avoid passing `RRC`'s by reference because they themselves represent a reference to something. Not sure if this has any profound impact on performance but more of a stylistic thing

rachitnigam commented on calyx-opt/src/passes/static_inliner.rs:
```diff
@@ -0,0 +1,388 @@
+                }
+                assert!(
+                    *latency == cur_offset,
+                    "static group latency doesn't match static seq latency"
```
I've left a question about this below as well but does the control operator itself need to specify the latency of things for any good reason if the only use is for debugging? It seem it would be more ergonomic if the control operators did not specify latencies of things and relied on the compiler to either infer that latency or simply never reify the latency up till this point where groups are inlined and therefore latency can be computed

rachitnigam commented on calyx-opt/src/passes/static_inliner.rs:
```diff
@@ -0,0 +1,388 @@
+                            let fgroup = StaticInliner::inline_static_control(
+                                fbranch, builder,
+                            );
+                            assert_eq!(fbranch_latency, fgroup.borrow().get_latency(), "tru branch and tru branch group latency do not match");
```
Incorrect error message

rachitnigam commented on calyx-opt/src/passes/static_inliner.rs:
```diff
@@ -0,0 +1,388 @@
+
+                // update trgoup_assigns to have guard %[0:tbranch_latency] in front of
+                // each assignment, and %[0:fbranch_latency] for fgroup_assigns
+                StaticInliner::update_assignment_timing(
```
Seems like the assignments on the `tgroup` and `fgroup` are iterated over twice. Once to add the timing guard and once to add the conditional guard. We can probably do this in one pass?

rachitnigam commented on calyx-opt/src/passes/static_inliner.rs:
```diff
@@ -0,0 +1,388 @@
+                latency,
+                attributes,
+            }) => {
+                let if_group = builder.add_static_group("static_if", *latency);
```
Is the `latency` of a `StaticIf` the max of the latencies of the two branches?

calebmkim replied:
yes.

rachitnigam commented on calyx-opt/src/passes/static_inliner.rs:
```diff
@@ -0,0 +1,388 @@
+        let control_ref = Rc::clone(&comp.control);
+        let mut builder = ir::Builder::new(comp, sigs);
+        // builds static map, which maps static islands to equivalent singular inlined static groups
+        self.build_static_map(&control_ref.borrow(), &mut builder);
```
The way this pass is structured, it traverses the control program twice. Once during `build_static_map` and another time in the visitor itself. Can we combine the two? Or rather, is there a reason that `build_static_map` needs to be wholly computed before traversing the program?

rachitnigam reviewed (commented):
Okay, this is super exciting! I have a couple of high-level thoughts:
1. Is it necessary for static control operators to specify their latencies? It seems to me that the latency is only used for validation. If that is the case, then we should probably remove it. In the currently world where all control operators specify their latencies, changing the latency of one static group might require meticulously changing the latency of many control operators which doesn't seem useful
2. The inlining pass is written so that it traverses the control program twice: once to build the "static map" and another time to replace the static islands. Is this necessary? Can we not simply replace the static islands while traversing the program?
3. I see what you meant about the guards being quite redundant. It seems we should be able to simplify them pretty easily in a future. This also makes me think if we can take dynamic guards like `f.out == 3` and abstract them into `%3` so that simplifying them becomes easier.

Let me know your thoughts on the comments!

calebmkim commented:
1.  Just changed the frontend so that latencies don't have to be explicitly given in the control. I still allow it optionally though. 
2. You're completely right, not sure why I was doing it like that. Fixed now. 
3. Yep. In fact, there's this one method on `Guard` called `add_interval(self, (beg, end))` that basically just changes `self -> self & (beg, end)`. I can probably change it to look thru `self` first to see if `(beg, end)` will be redundant, and then just not change anything if it will obviously be redundant. 

Also, I'll look thru the minor reviews u made and update accordingly as well.

EclecticGriffin commented:
Making a strong argument for hashconsing the guard expressions!

EclecticGriffin commented on calyx-ir/src/control.rs:
```diff
@@ -215,6 +215,16 @@ impl GetAttributes for StaticEnable {
+    where
+        S: Into<Id>,
+    {
+        self.get_attributes().get(attr).cloned()
```
Prefer `copied` over `cloned` here since `u64` is a copy type.

EclecticGriffin commented on calyx-ir/src/control.rs:
```diff
@@ -477,6 +479,11 @@ impl Control {
     }
 
+    /// Convience constructor for a static control
+    pub fn static_control(sc: StaticControl) -> Self {
+        Control::Static(sc)
+    }
```
Probably worthwhile to just implement `From<StaticControl>` for `Control` and
use that instead

EclecticGriffin commented on calyx-ir/src/control.rs:
```diff
@@ -500,6 +507,62 @@ impl StaticControl {
     }
 
+    /// Convience constructor for static enable.
+    pub fn enable(group: RRC<StaticGroup>) -> Self {
+        StaticControl::Enable(StaticEnable {
+            group,
+            attributes: Attributes::default(),
+        })
```
This one can also be a `From` implementation

EclecticGriffin commented on calyx-ir/src/control.rs:
```diff
@@ -500,6 +507,62 @@ impl StaticControl {
+    }
+
+    /// Convience constructor for static if
+    pub fn if_(
```
nit: I don't personally love the trailing underscore, maybe something like
`static_if` or `if_static`

calebmkim commented on calyx-frontend/src/ast.rs:
```diff
@@ -341,6 +343,54 @@ pub enum Control {
+        /// Attributes
+        attributes: Attributes,
+        /// Latency for the seq
+        latency: Option<u64>,
```
Interesting... the only question is, if someone _does_ annotate with latency of 0, do we want to have the latency as 0, then throw an error when it doesn't match the latency that we infer? If we use NonZerou64, it seems like it will just act as if the (incorrect) 0 latency annotation is not there-- which seems fine too. Ig it's a preference.

rachitnigam replied:
We can probably do that in the parser itself using the `input.error` message and reject that program before it appears in the AST?

rachitnigam commented:
Another drive-by thought: worth making sure that the following two cases work:
1. Loop with 1-cycle body
2. Doubly nested loop with 1-cycle body

I'll come by and take another look but if you're confident things work, then merge it! Once #1419 is merged, we should run the inlining pass on the Dahlia benchmarks from the evaluation repository since those are a good test case for all of this stuff. Alternatively, if you're feeling adventurous, you can change the Dahlia frontend to generate `static group` instead of the annotated stuff (on a branch) and see if any of the polybench benchmarks fail

calebmkim commented:
Ok, I added a test case for nested loop w/ one cycle body! I think I'm gonna merge since it'll be easier for @paili0628 to make progress on some of her work if we merge sooner rather than later.
